### DISCUSSION TITLE
Is The WebGL2 Backend Production Ready?

### DISCUSSION THREAD

**Discussion by zicklag** [gfx-rs/wgpu#3003] [Q&A]
_Originally posted by @andrewvarga here: https://github.com/gfx-rs/wgpu/issues/1617#issuecomment-1230965602_

> Hi all! I asked this on the Wgpu Users channel but maybe it's better to track on github:
https://matrix.to/#/!XFRnMvAfptAHthwBCx:matrix.org/$xqMiU6EZP9Vk4K1Rl7VDxmXiXeeAvqdcIYeau80gNSY?via=matrix.org&via=mozilla.org&via=kyju.org
>
> In short, I'm just curious if the WebGL2 backend is considered stable enough for production at this point, and if it is maybe I'm doing something wrong?
> Something as simple as using a uniform buffer to set the color of a triangle seems not to work when target is WebGL2, and the official examples have problems on my end.

---

Hi @andrewvarga!

I can't get to matrix, so I can't get to your exact question, but I think that the WebGL2 backend is fairly "production-ish" ready, with the caveat that WebGL2 will never be able to be a fully-compliant WGPU implementation. That doesn't mean you can't use it and get great results, though.

For instance, here's a Bevy WebGL2 demo with PBR: https://bevyengine.org/examples/3d/load-gltf/.

As far as the official examples, both the mipmap and shadow examples on the official site don't work right for me, but the other ones do.

I think shadow should be fixable, but I don't know what's causing the issue. I don't know that mipmap is ever going to work right or not, I can't remember if there's a way to get WebGL to run that correctly.

As far as setting the color of a triangle, you should _absolutely_ be able to get that working. Does the Bevy example linked to above work for you. And how about the WGPU.rs [Cube example](https://wgpu.rs/examples-gl/?example=cube)?

If neither of those work, it might be a driver/browser/hardware issue.

**Comment by andrewvarga** [comment]
Hey @zicklag, thanks for getting back on this, the Bevy example looks really promising.

I've created a simple branch with my changes which modifies the cube example by adding a uniform buffer for the color:
https://github.com/andrewvarga/wgpu/pull/1/files

When running this with WebGPU I'm getting a red triangle, but it is white on WebGL2. It's possible that I did an easy oversight somewhere.. would be interesting to see if you see the same thing, I'm on an M1 Macbook but I can test this out on Windows too.

**Reply by zicklag** [reply]
Just pulled and ran your branch and it works for me with WebGL2:

![image](https://user-images.githubusercontent.com/25393315/187480409-23c67243-d73d-4087-81dd-0e18e88f625d.png)

What browser are you using?

**Reply by andrewvarga** [reply]
> Just pulled and ran your branch and it works for me with WebGL2:
> 
> ![image](https://user-images.githubusercontent.com/25393315/187480409-23c67243-d73d-4087-81dd-0e18e88f625d.png)
> 
> What browser are you using?

How can I get that info to show? Don't see it in the browser console.

I'm using the latest Chrome on mac: Version 104.0.5112.101 (Official Build) (arm64), but getting the same result on Firefox and Safari.

**Reply by zicklag** [reply]
That was in my browser console. Strange that you don't see it. :thinking: 

![image](https://user-images.githubusercontent.com/25393315/187482428-a9c6c224-709c-47fc-b360-c1b1ab8cfc4a.png)

**Reply by andrewvarga** [reply]
oh I see! I think you're running the `hello-triangle` example, and my update was for the `cube` example, so I'm using this command:
`cargo run-wasm --example cube --features webgl`

Running the triangle example, I'm also seeing those logs:
```
Vendor: Google Inc. (Apple)
hello-triangle.js:825 Renderer: ANGLE (Apple, Apple M1 Max, OpenGL 4.1)
hello-triangle.js:825 Version: WebGL 2.0 (OpenGL ES 3.0 Chromium)
hello-triangle.js:825 SL version: WebGL GLSL ES 3.00 (OpenGL ES GLSL ES 3.0 Chromium)
hello-triangle.js:825 Adapter Gl AdapterInfo { name: "ANGLE (Apple, Apple M1 Max, OpenGL 4.1)", vendor: 0, device: 0, device_type: IntegratedGpu, backend: Gl }
hello-triangle.js:825 	Compiled shader WebShaderKey(2v3)
hello-triangle.js:825 	Compiled shader WebShaderKey(1v3)
hello-triangle.js:825 	Linked program WebProgramKey(2v1)
hello-triangle.js:825 Created render pipeline Valid((0, 1, Gl)) with RenderPipelineDescriptor { label: None, layout: Some((0, 1, Gl)), vertex: VertexState { stage: ProgrammableStageDescriptor { module: (0, 1, Gl), entry_point: "vs_main" }, buffers: [] }, primitive: PrimitiveState { topology: TriangleList, strip_index_format: None, front_face: Ccw, cull_mode: None, unclipped_depth: false, polygon_mode: Fill, conservative: false }, depth_stencil: None, multisample: MultisampleState { count: 1, mask: 18446744073709551615, alpha_to_coverage_enabled: false }, fragment: Some(FragmentState { stage: ProgrammableStageDescriptor { module: (0, 1, Gl), entry_point: "fs_main" }, targets: [Some(ColorTargetState { format: Rgba8UnormSrgb, blend: None, write_mask: RED | GREEN | BLUE | ALPHA | COLOR | ALL })] }), multiview: None }
hello-triangle.js:825 configuring surface with SurfaceConfiguration { usage: RENDER_ATTACHMENT, format: Rgba8UnormSrgb, width: 2048, height: 1536, present_mode: Fifo }
```

**Reply by zicklag** [reply]
Whoops! You're right, I'm running the wrong example. And I can confirm the white cube:

![image](https://user-images.githubusercontent.com/25393315/187484126-688e3d33-84d0-406d-97ac-387b6549d3db.png)

I'll look at that a little closer.

**Reply by zicklag** [reply]
A quick check with Spector.js shows that it's not actually rendering "white' it's rendering transparent: 

![image](https://user-images.githubusercontent.com/25393315/187484917-46f8c7b7-d18d-46e5-83a3-587797451794.png)

Not sure what that means yet.

**Reply by zicklag** [reply]
Ah, looks like the uniform value isn't getting set properly. The alpha is negative and the other values don't make much sense either. Something must be going wrong while uploading the buffer value.

![image](https://user-images.githubusercontent.com/25393315/187485389-cd023a81-de53-4a13-a0e3-c7d52e42cc50.png)

**Reply by andrewvarga** [reply]
that's interesting..and yes I saw similar results in my own prototype where the buffer value uploaded didn't match the value I expected..

**Reply by zicklag** [reply]
I did a little bit of messing around, so far with no luck discovering the issue. I even tried writing to the buffer every frame with no luck. Seems like it doesn't matter what we put in it.

This is such a simple example, I know I've done similar stuff and had it work, so maybe we're just missing something obvious.

I'm out of time to look at it for now, but I'll come back to it next I get the chance and see if I can figure anything out.

**Reply by andrewvarga** [reply]
thanks!! Only thing I was thinking of was this line with the min_binding_size, which I'm not really sure about, or maybe some padding requirement..
`min_binding_size: None,`
https://github.com/andrewvarga/wgpu/blob/404ee7264d2df345b59f74cb7db796ccc80bb9a5/wgpu/examples/cube/main.rs#L188

**Reply by zicklag** [reply]
Yeah, I already tried that, and unfortunately it doesn't fix it.

If you set that value, WGPU is supposed to error if you pass in a buffer that's too small. For instance, in this case it makes sense to set the min length to 16 ( the number of bytes in 4 f32's ), but you don't have to ( and it doesn't fix our current issue if you do ).

If you set the min length to 16, then WGPU would throw an error if at some point you accidentally bound it to a buffer that only had 2 f32's in it, so it can be useful for preventing mistakes.

**Reply by andrewvarga** [reply]
@zicklag  Do you think it can be related to `bytemuck::cast_slice`?
If you can point out the area of code which is reponsible for this in the webgl2 backend, I can also try to investigate this.

**Reply by zicklag** [reply]
> @zicklag Do you think it can be related to bytemuck::cast_slice?

I highly doubt it. Bytemuck has been used widely all over the place I think and I can't imagine that it's having an issue casting 4 f32's properly in this case.

I believe [this](https://github1s.com/gfx-rs/wgpu/blob/HEAD/wgpu-hal/src/gles/device.rs#L345-L346) is the code responsible for uploading buffers on WebGL.

Note that the code is shared with the OpenGL backend, which actually works and displays the color correctly, so the issue may have something to do with nuances specific to WebGL, but in code shared by WebGL and OpenGL.

**Reply by zicklag** [reply]
I just changed the bindings, so that the projection matrix was at binding 2 and the color buffer was at binding 0 and it works on native, but on WASM I get a different error:

![image](https://user-images.githubusercontent.com/25393315/188013722-a17c1da0-b56c-467c-b94d-f2717e19c3d5.png)

Not sure how that helps us yet, but it's something different anyway. :man_shrugging:

**Reply by zicklag** [reply]
I found something: both uniforms are using the same buffer on WebGL:

![image](https://user-images.githubusercontent.com/25393315/188014598-dcbadae8-7b00-4cd6-93eb-eebb26a16716.png)

So our color uniform is actually grabbing the first 4 floats from the projection matrix, which is obviously wrong.

**Reply by zicklag** [reply]
This feels like a bug. As far as I can tell so far it seems that the commands issued by WGPU _seem_ right:

![image](https://user-images.githubusercontent.com/25393315/188017682-c4f829ff-51d8-41d0-8148-090cbc0ef23c.png)

But notice that we are making two uniform block bindings, but there is only _one_ uniform block in the output fragment shader:

```glsl
#version 300 es

precision highp float;
precision highp int;
struct VertexOutput {
    vec2 tex_coord;
    vec4 position;
};
// The ONLY uniform block in this fragment shader.
// That could be why this is referencing the uniform buffer 0 instead of uniform buffer 1
// like it should be..
uniform type_1_block_0Fragment {
    vec4 _group_0_binding_2_fs;
};
smooth in vec2 _vs2fs_location0;
layout(location = 0) out vec4 _fs2p_location0;
void main() {
    VertexOutput vertex = VertexOutput(_vs2fs_location0, gl_FragCoord);
    vec4 _e4 = _group_0_binding_2_fs;
    _fs2p_location0 = _e4;
    return;
}
```

I'm wondering if `naga` needs to output the unused projection matrix uniform block in the fragment shader just so that it the index ends up correct.

**Edit:** Nevermind about naga needing to output the other uniform block, I tested that's not the issue. I have to find out where we tie the uniform block buffers to the block names ( if that's how it works, I haven't used uniform blocks in WebGL yet ).

**Reply by zicklag** [reply]
It looks like it's not generating the uniform block bindings. Looks like it might be an issue during the shader reflection [here](https://github1s.com/gfx-rs/wgpu/blob/HEAD/wgpu-hal/src/gles/device.rs#L26-L27).

Because we're never hitting this [line](https://github1s.com/gfx-rs/wgpu/blob/HEAD/wgpu-hal/src/gles/device.rs#L277-L278), which I think is necessary to map the buffer binding to the block in the shader.

**Reply by zicklag** [reply]
It's not finding the uniform name [here](https://github1s.com/gfx-rs/wgpu/blob/HEAD/wgpu-hal/src/gles/device.rs#L47-L48), so it's not generating the name_binding_map entry.

**Reply by zicklag** [reply]
@cwfitzgerald if you get the chance to look a this, do think that naga should be finding the uniform name here? Do you think I'm on the right track with this?

**Reply by andrewvarga** [reply]
I created a github issue to track this: https://github.com/gfx-rs/wgpu/issues/3024
In the meantime I guess a workaround would be to use a single bind group instead of two (?)

**Reply by zicklag** [reply]
Good idea to open an issue.

> In the meantime I guess a workaround would be to use a single bind group instead of two (?)

I'm not super familiar with the terminology, but I think we are already using a single bind group, but different bindings? ( Because both use `@group(0)`, but different `@binding([number])` )

But I know what you mean, and it seems worth a try.

It might also be worth trying to put it in a separate group. I was just looking at the Bevy [sprite shader](https://github.com/bevyengine/bevy/blob/main/crates/bevy_sprite/src/render/sprite.wgsl) and it looks like it uses a separate group for the fragment shader data. That might be something worth trying.

**Comment by ryankaplan** [comment]
@zicklag @andrewvarga do either of you know if this has been fixed? I'm using the WebGL backend and I'm running into an issue that looks a lot like shaders getting garbage uniform data. I'm wondering if it's the same problem you ran into.

I'm asking because it's been a year and I know folks have been working on the WebGL backend. So I wanted to check if this is stale before digging further.

**Reply by cwfitzgerald** [reply]
Take a look at https://github.com/gfx-rs/wgpu/issues/3674, this should have the majority of the known issues
